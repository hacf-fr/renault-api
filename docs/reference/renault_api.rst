Renault high-level API Reference
================================

.. contents::
    :local:
    :backlinks: none


Session
---------------

.. automodule:: renault_api.renault_session
   :members:

Client
--------------

.. automodule:: renault_api.renault_client
   :members:

Account
---------------

.. automodule:: renault_api.renault_account
   :members:

Vehicle
---------------

.. automodule:: renault_api.renault_vehicle
   :members:
