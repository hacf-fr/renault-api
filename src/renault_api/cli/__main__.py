"""Command-line interface."""
import errno
import json
import logging
import os
from datetime import datetime
from io import TextIOWrapper
from typing import Any
from typing import Dict
from typing import Optional

import aiohttp
import click
from click.core import Context

from . import helpers
from . import renault_account
from . import renault_client
from . import renault_settings
from . import renault_vehicle
from .charge import commands as charge_commands
from .hvac import commands as hvac_commands
from renault_api.credential_store import FileCredentialStore


_WARNING_DEBUG_ENABLED = (
    "Debug output enabled. Logs may contain personally identifiable "
    "information and account credentials! Be sure to sanitise these logs "
    "before sending them to a third party or posting them online."
)


def _check_for_debug(debug: bool, log: bool) -> None:
    """Renault CLI."""
    if debug or log:
        renault_log = logging.getLogger("renault_api")
        renault_log.setLevel(logging.DEBUG)

        if log:
            # create directory
            try:
                os.makedirs("logs")
            except OSError as e:  # pragma: no cover
                if e.errno != errno.EEXIST:
                    raise

            # create formatter and add it to the handlers
            formatter = logging.Formatter(
                "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
            )

            # create file handler which logs even debug messages
            fh = logging.FileHandler(f"logs/{datetime.today():%Y-%m-%d}.log")
            fh.setLevel(logging.DEBUG)
            fh.setFormatter(formatter)

            # And enable our own debug logging
            renault_log.addHandler(fh)

        if debug:
            logging.basicConfig()

        renault_log.warning(_WARNING_DEBUG_ENABLED)


@click.group()
@click.version_option()
@click.option("--debug", is_flag=True, help="Display debug traces.")
@click.option("--log", is_flag=True, help="Log debug traces to file.")
@click.option("--json", is_flag=True, help="Return data as JSON")
@click.option("--locale", default=None, help="API locale (eg. fr_FR)")
@click.option(
    "--account",
    default=None,
    help="Kamereon account ID to use",
)
@click.option("--vin", default=None, help="Vehicle VIN to use")
@click.pass_context
def main(
    ctx: Context,
    *,
    debug: bool,
    log: bool,
    json: bool,
    locale: Optional[str] = None,
    account: Optional[str] = None,
    vin: Optional[str] = None,
) -> None:
    """Main entry point for the Renault CLI."""
    ctx.ensure_object(dict)
    ctx.obj["credential_store"] = FileCredentialStore(
        os.path.expanduser(renault_settings.CREDENTIAL_PATH)
    )
    _check_for_debug(debug, log)
    ctx.obj["json"] = json
    if locale:
        ctx.obj["locale"] = locale
    if account:
        ctx.obj["account"] = account
    if vin:  # pragma: no branch
        ctx.obj["vin"] = vin


main.add_command(charge_commands.charge)
main.add_command(hvac_commands.hvac)


@main.command()
@click.pass_obj
@helpers.coro_with_websession
async def accounts(
    ctx_data: Dict[str, Any],
    *,
    websession: aiohttp.ClientSession,
) -> None:
    """Display list of accounts."""
    await renault_client.display_accounts(websession, ctx_data)


@main.command()
@click.option("--user", prompt=True)
@click.option("--password", prompt=True, hide_input=True)
@click.pass_obj
@helpers.coro_with_websession
async def login(
    ctx_data: Dict[str, Any],
    *,
    user: str,
    password: str,
    websession: aiohttp.ClientSession,
) -> None:
    """Login to Renault."""
    await renault_client.login(websession, ctx_data, user, password)


@main.command()
def reset() -> None:
    """Clear all credentials/settings from the credential store."""
    renault_settings.reset()


@main.command()
@click.option("--locale", default=None, help="API locale (eg. fr_FR)")
@click.option(
    "--account", default=None, help="Kamereon account ID to use for future calls"
)
@click.option("--vin", default=None, help="Vehicle VIN to use for future calls")
@click.pass_obj
@helpers.coro_with_websession
async def set(
    ctx_data: Dict[str, Any],
    *,
    locale: Optional[str] = None,
    account: Optional[str] = None,
    vin: Optional[str] = None,
    websession: aiohttp.ClientSession,
) -> None:
    """Store specified settings into credential store."""
    await renault_settings.set_options(websession, ctx_data, locale, account, vin)


@main.command()
@click.pass_obj
def settings(ctx_data: Dict[str, Any]) -> None:
    """Display the current configuration keys."""
    renault_settings.display_settings(ctx_data)


@main.command()
@click.pass_obj
@helpers.coro_with_websession
async def status(
    ctx_data: Dict[str, Any],
    *,
    websession: aiohttp.ClientSession,
) -> None:
    """Display vehicle status."""
    await renault_vehicle.display_status(websession, ctx_data)


@main.command()
@click.pass_obj
@helpers.coro_with_websession
async def vehicles(
    ctx_data: Dict[str, Any],
    *,
    websession: aiohttp.ClientSession,
) -> None:
    """Display list of vehicles."""
    await renault_account.display_vehicles(websession, ctx_data)


@main.command()
@click.pass_obj
@helpers.coro_with_websession
async def vehicle(
    ctx_data: Dict[str, Any],
    *,
    websession: aiohttp.ClientSession,
) -> None:
    """Display vehicle details."""
    await renault_vehicle.display_vehicle(websession, ctx_data)


@main.command()
@click.pass_obj
@helpers.coro_with_websession
async def contracts(
    ctx_data: Dict[str, Any],
    *,
    websession: aiohttp.ClientSession,
) -> None:
    """Display vehicle contracts."""
    await renault_vehicle.display_contracts(websession, ctx_data)


@main.group()
def http() -> None:
    """Raw HTTP."""
    pass


@http.command(name="get")
@click.argument("endpoint")
@click.pass_obj
@helpers.coro_with_websession
async def http_get(
    ctx_data: Dict[str, Any],
    *,
    endpoint: str,
    websession: aiohttp.ClientSession,
) -> None:
    """Process HTTP GET request on endpoint."""
    await renault_client.http_request(websession, ctx_data, "GET", endpoint)


@http.command(name="post-file")
@click.argument("endpoint")
@click.argument("json-body", type=click.File("rb"))
@click.pass_obj
@helpers.coro_with_websession
async def http_post_file(
    ctx_data: Dict[str, Any],
    *,
    endpoint: str,
    json_body: TextIOWrapper,
    websession: aiohttp.ClientSession,
) -> None:
    """Process HTTP POST request on endpoint."""
    await renault_client.http_request(
        websession, ctx_data, "POST", endpoint, json.load(json_body)
    )


@http.command(name="post")
@click.argument("endpoint")
@click.argument("json-body")
@click.pass_obj
@helpers.coro_with_websession
async def http_post(
    ctx_data: Dict[str, Any],
    *,
    endpoint: str,
    json_body: str,
    websession: aiohttp.ClientSession,
) -> None:
    """Process HTTP POST request on endpoint."""
    await renault_client.http_request(
        websession, ctx_data, "POST", endpoint, json.loads(json_body)
    )


if __name__ == "__main__":  # pragma: no cover
    main(prog_name="renault-api")
