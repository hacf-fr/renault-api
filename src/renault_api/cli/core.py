"""Core functions for the CLI."""
import logging
import os
from datetime import datetime
from locale import getdefaultlocale
from typing import Optional

import click

from renault_api.const import CONF_LOCALE, PERMANENT_KEYS
from renault_api.credential_store import CredentialStore
from renault_api.credential_store import FileCredentialStore
from renault_api.exceptions import RenaultException
from renault_api.helpers import get_api_keys
from renault_api.model.credential import Credential

CONF_ACCOUNT_ID = "accound-id"
CONF_VIN = "vin"


class CLICredentialStore:
    """Singleton of the CredentialStore for the CLI."""

    __instance: CredentialStore = None

    def __new__(cls):
        """Get singleton Credential Store."""
        if not CLICredentialStore.__instance:
            CLICredentialStore.__instance = FileCredentialStore(
                os.path.expanduser("~/.credentials/renault-api.json")
            )
        return CLICredentialStore.__instance


def set_debug(debug: bool, log: bool) -> None:
    """Renault CLI."""
    if debug or log:
        renault_log = logging.getLogger("renault_api")
        renault_log.setLevel(logging.DEBUG)

        if log:
            # create formatter and add it to the handlers
            formatter = logging.Formatter(
                "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
            )

            # create file handler which logs even debug messages
            fh = logging.FileHandler(f"logs/{datetime.today():%Y-%m-%d}.log")
            fh.setLevel(logging.DEBUG)
            fh.setFormatter(formatter)

            # And enable our own debug logging
            renault_log.addHandler(fh)

        if debug:
            logging.basicConfig()

        renault_log.warning(
            "Debug output enabled. Logs may contain personally identifiable "
            "information and account credentials! Be sure to sanitise these logs "
            "before sending them to a third party or posting them online."
        )


async def set_options(
    websession, locale: Optional[str], account: Optional[str], vin: Optional[str]
) -> None:
    """Set configuration keys."""
    await set_options(locale, account, vin)
    if locale:
        await set_locale(websession, locale)
    if account:
        credential_store = CLICredentialStore()
        credential_store[CONF_ACCOUNT_ID] = Credential(account)
    if vin:
        credential_store = CLICredentialStore()
        credential_store[CONF_VIN] = Credential(vin)


async def ensure_locale(websession) -> None:
    """Prompt the user for locale."""
    credential_store = CLICredentialStore()
    if CONF_LOCALE in credential_store:
        return

    default_locale = getdefaultlocale()[0]
    while True:
        locale = click.prompt("Please select a locale", default=default_locale)
        try:
            await set_locale(websession, locale)
        except RenaultException as exc:
            click.echo(str(exc), err=True)
        else:
            return
        click.echo(f"Locale `{locale}` is unknown.", err=True)


async def set_locale(websession, locale: str) -> None:
    """Prompt the user for locale."""
    await get_api_keys(locale, aiohttp_session=websession)

    credential_store = CLICredentialStore()
    credential_store[CONF_LOCALE] = Credential(locale)


async def display_keys(websession) -> None:
    """Get the current configuration keys."""
    await ensure_locale(websession)
    credential_store = CLICredentialStore()
    for key in PERMANENT_KEYS:
        click.echo(f"Current {key}: {credential_store.get_value(key)}")
