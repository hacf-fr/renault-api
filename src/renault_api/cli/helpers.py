"""Helpers for Renault API."""
import asyncio
import functools
from datetime import datetime
from datetime import timedelta
from typing import Any
from typing import Callable
from typing import Optional
from typing import Tuple

import aiohttp
import click
import dateparser
import tzlocal

from renault_api.exceptions import RenaultException
from renault_api.kamereon.helpers import DAYS_OF_WEEK


_DATETIME_FORMAT = "%Y-%m-%d %H:%M:%S"


def coro_with_websession(func: Callable[..., Any]) -> Callable[..., Any]:
    """Ensure the routine runs on an event loop with a websession."""

    async def run_command(func: Callable[..., Any], *args: Any, **kwargs: Any) -> None:
        async with aiohttp.ClientSession() as websession:
            try:
                kwargs["websession"] = websession
                await func(*args, **kwargs)
            except RenaultException as exc:  # pragma: no cover
                raise click.ClickException(str(exc)) from exc
            finally:
                closed_event = create_aiohttp_closed_event(websession)
                await websession.close()
                await closed_event.wait()

    def wrapper(*args: Any, **kwargs: Any) -> None:
        asyncio.run(run_command(func, *args, **kwargs))

    return functools.update_wrapper(wrapper, func)


def days_of_week_option(helptext: str) -> Callable[..., Any]:
    """Add day of week string options."""

    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
        for day in reversed(DAYS_OF_WEEK):
            func = click.option(
                f"--{day}",
                help=helptext.format(day.capitalize()),
            )(func)
        return func

    return decorator


def start_end_option(add_period: bool) -> Callable[..., Any]:
    """Add start/end options."""

    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
        func = click.option(
            "--from", "start", help="Date to start showing history from", required=True
        )(func)
        func = click.option(
            "--to",
            "end",
            help="Date to finish showing history at (cannot be in the future)",
            required=True,
        )(func)
        if add_period:
            func = click.option(
                "--period",
                default="month",
                help="Period over which to aggregate.",
                type=click.Choice(["day", "month"], case_sensitive=False),
            )(func)
        return func

    return decorator


def create_aiohttp_closed_event(
    websession: aiohttp.ClientSession,
) -> asyncio.Event:  # pragma: no cover
    """Work around aiohttp issue that doesn't properly close transports on exit.

    See https://github.com/aio-libs/aiohttp/issues/1925#issuecomment-639080209

    Args:
        websession (aiohttp.ClientSession): session for which to generate the event.

    Returns:
        An event that will be set once all transports have been properly closed.
    """
    transports = 0
    all_is_lost = asyncio.Event()

    def connection_lost(exc, orig_lost):  # type: ignore
        nonlocal transports

        try:
            orig_lost(exc)
        finally:
            transports -= 1
            if transports == 0:
                all_is_lost.set()

    def eof_received(orig_eof_received):  # type: ignore
        try:
            orig_eof_received()
        except AttributeError:
            # It may happen that eof_received() is called after
            # _app_protocol and _transport are set to None.
            pass

    for conn in websession.connector._conns.values():  # type: ignore
        for handler, _ in conn:
            proto = getattr(handler.transport, "_ssl_protocol", None)
            if proto is None:
                continue

            transports += 1
            orig_lost = proto.connection_lost
            orig_eof_received = proto.eof_received

            proto.connection_lost = functools.partial(
                connection_lost, orig_lost=orig_lost
            )
            proto.eof_received = functools.partial(
                eof_received, orig_eof_received=orig_eof_received
            )

    if transports == 0:
        all_is_lost.set()

    return all_is_lost


def parse_dates(start: str, end: str) -> Tuple[datetime, datetime]:
    """Convert start/end string arguments into datetime arguments."""
    parsed_start = dateparser.parse(start)
    parsed_end = dateparser.parse(end)

    if not parsed_start:  # pragma: no cover
        raise ValueError(f"Unable to parse `{start}` into start datetime.")
    if not parsed_end:  # pragma: no cover
        raise ValueError(f"Unable to parse `{end}` into end datetime.")

    return (parsed_start, parsed_end)


def _timezone_offset() -> int:
    """Return UTC offset in minutes."""
    utcoffset = tzlocal.get_localzone().utcoffset(datetime.now())
    if utcoffset:
        return int(utcoffset.total_seconds() / 60)
    return 0  # pragma: no cover


def _format_tzdatetime(date_string: str) -> str:
    date = datetime.fromisoformat(date_string.replace("Z", "+00:00"))
    return str(date.astimezone(tzlocal.get_localzone()).strftime(_DATETIME_FORMAT))


def _format_tztime(time: str) -> str:
    total_minutes = int(time[1:3]) * 60 + int(time[4:6]) + _timezone_offset()
    hours, minutes = divmod(total_minutes, 60)
    hours = hours % 24  # Ensure it is 00-23
    return f"{hours:02g}:{minutes:02g}"


def convert_minutes_to_tztime(minutes: int) -> str:
    """Convert minutes to Thh:mmZ format."""
    total_minutes = minutes - _timezone_offset()
    hours, minutes = divmod(total_minutes, 60)
    hours = hours % 24  # Ensure it is 00-23
    return f"T{hours:02g}:{minutes:02g}Z"


def _format_minutes(mins: float) -> str:
    d = timedelta(minutes=mins)
    return str(d)


def get_display_value(
    value: Optional[Any] = None,
    unit: Optional[str] = None,
) -> str:
    """Get a display for value."""
    if value is None:  # pragma: no cover
        return ""
    if unit is None:
        return str(value)
    if unit == "tzdatetime":
        return _format_tzdatetime(value)
    if unit == "tztime":
        return _format_tztime(value)
    if unit == "minutes":
        return _format_minutes(value)
    if unit == "kW":
        value = value / 1000
        return f"{value:.2f} {unit}"
    return f"{value} {unit}"
