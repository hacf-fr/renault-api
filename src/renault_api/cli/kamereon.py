"""Singletons for the CLI."""
import click
from tabulate import tabulate

from .core import CLICredentialStore
from .core import CONF_ACCOUNT_ID
from .core import CONF_VIN
from .core import ensure_locale
from renault_api.const import CONF_LOCALE
from renault_api.exceptions import RenaultException
from renault_api.helpers import get_api_keys
from renault_api.kamereon import CREDENTIAL_GIGYA_LOGIN_TOKEN
from renault_api.kamereon import Kamereon


class CLIKamereon:
    """Singleton of the CredentialStore for the CLI."""

    __instance: Kamereon = None

    @classmethod
    async def get_client(cls, websession):
        """Get singleton Kamereon client."""
        if not CLIKamereon.__instance:
            await ensure_locale(websession)

            credential_store = CLICredentialStore()
            locale = credential_store.get_value(CONF_LOCALE)
            country = locale[-2:]
            api_keys = await get_api_keys(locale, aiohttp_session=websession)
            return Kamereon(
                websession, country, api_keys, credential_store=credential_store
            )
        return CLIKamereon.__instance


async def ensure_logged_in(websession) -> None:
    """Prompt the user for credentials."""
    await ensure_locale(websession)

    credential_store = CLICredentialStore()
    if CREDENTIAL_GIGYA_LOGIN_TOKEN in credential_store:
        return

    while True:
        user = click.prompt("user")
        password = click.prompt("password", hide_input=True)
        try:
            await do_login(websession, user, password)
        except RenaultException as exc:
            click.echo(f"Login failed: {exc}.", err=True)
        else:
            return


async def do_login(websession, user, password):
    """Attempt login."""
    await ensure_locale(websession)
    kamereon = await CLIKamereon.get_client(websession)
    await kamereon.login(user, password)


async def get_account(websession) -> str:
    """Prompt the user for account."""
    credential_store = CLICredentialStore()
    if CONF_ACCOUNT_ID in credential_store:
        return credential_store.get_value(CONF_ACCOUNT_ID)

    await ensure_logged_in(websession)
    kamereon = await CLIKamereon.get_client(websession)
    response = await kamereon.get_person()
    if not response.accounts:
        raise RenaultException("No account found.")
    if len(response.accounts) == 1:
        return response.accounts[0].accountId
    elif len(response.accounts) > 1:
        menu = "Multiple accounts found:\n"
        for i, account in enumerate(response.accounts):
            menu = menu + \
                f"\t[{i+1}] {account.accountId} ({account.accountType})\n"

        while True:
            i = int(click.prompt(f"{menu}Please select"))
            try:
                account = response.accounts[i - 1].accountId
            except (KeyError, IndexError) as exc:
                click.echo(f"Invalid option: {exc}.", err=True)
            else:
                click.echo("To avoid seeing this message again, use:"
                           f"> renault-api set --account {account}")
                return account


async def get_vin(websession, account):
    """Prompt the user for vin."""
    credential_store = CLICredentialStore()
    if CONF_VIN in credential_store:
        return

    await ensure_logged_in(websession)
    kamereon = await CLIKamereon.get_client(websession)
    response = await kamereon.get_vehicles(account)
    if not response.vehicleLinks:
        raise RenaultException("No vehicle found.")
    if len(response.vehicleLinks) == 1:
        return response.vehicleLinks[0].vin
    elif len(response.vehicleLinks) > 1:
        menu = "Multiple vehicles found:\n"
        for i, vehicle in enumerate(response.vehicleLinks):
            menu = menu + f"\t[{i+1}] {vehicle.vin}\n"

        while True:
            i = int(click.prompt(f"{menu}Please select"))
            try:
                return response.vehicleLinks[i - 1].vin
            except (KeyError, IndexError) as exc:
                click.echo(f"Invalid option: {exc}.", err=True)


async def display_accounts(websession) -> None:
    """Login to Renault."""
    await ensure_logged_in(websession)
    kamereon = await CLIKamereon.get_client(websession)
    response = await kamereon.get_person()
    accounts = {
        account.accountType: account.accountId for account in response.accounts}
    click.echo(tabulate(accounts.items(), headers=[
               "Type", "ID"]))
