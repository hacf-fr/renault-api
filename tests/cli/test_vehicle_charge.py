"""Test cases for the __main__ module."""
from aioresponses import aioresponses
from aioresponses.core import RequestCall  # type:ignore
from click.testing import CliRunner
from tests import fixtures
from yarl import URL

from . import initialise_credential_store
from renault_api.cli import __main__


def test_charge_history_day(
    mocked_responses: aioresponses, cli_runner: CliRunner
) -> None:
    """It exits with a status code of zero."""
    initialise_credential_store(include_account_id=True, include_vin=True)
    fixtures.inject_get_charge_history(
        mocked_responses, start="20201101", end="20201130", period="day"
    )

    result = cli_runner.invoke(
        __main__.main, "charge history --from 2020-11-01 --to 2020-11-30 --period day"
    )
    assert result.exit_code == 0, result.exception

    expected_output = (
        "     Day    Number of charges  Total time charging      Errors\n"
        "--------  -------------------  ---------------------  --------\n"
        "20201208                    2  8:15:00                       0\n"
        "20201205                    1  10:57:00                      0\n"
    )
    assert expected_output == result.output


def test_charge_history_month(
    mocked_responses: aioresponses, cli_runner: CliRunner
) -> None:
    """It exits with a status code of zero."""
    initialise_credential_store(include_account_id=True, include_vin=True)
    fixtures.inject_get_charge_history(
        mocked_responses, start="202011", end="202011", period="month"
    )

    result = cli_runner.invoke(
        __main__.main, "charge history --from 2020-11-01 --to 2020-11-30"
    )
    assert result.exit_code == 0, result.exception

    expected_output = (
        "  Month    Number of charges  Total time charging      Errors\n"
        "-------  -------------------  ---------------------  --------\n"
        " 202011                    1  7:59:00                       0\n"
    )
    assert expected_output == result.output


def test_charge_mode_get(mocked_responses: aioresponses, cli_runner: CliRunner) -> None:
    """It exits with a status code of zero."""
    initialise_credential_store(include_account_id=True, include_vin=True)
    fixtures.inject_get_charge_mode(mocked_responses)

    result = cli_runner.invoke(__main__.main, "charge mode")
    assert result.exit_code == 0, result.exception

    expected_output = "Charge mode: always\n"
    assert expected_output == result.output


def test_charge_mode_set(mocked_responses: aioresponses, cli_runner: CliRunner) -> None:
    """It exits with a status code of zero."""
    initialise_credential_store(include_account_id=True, include_vin=True)
    url = fixtures.inject_set_charge_mode(mocked_responses, mode="schedule_mode")

    result = cli_runner.invoke(__main__.main, "charge mode --set schedule_mode")
    assert result.exit_code == 0, result.exception

    expected_json = {
        "data": {"attributes": {"action": "schedule_mode"}, "type": "ChargeMode"}
    }
    expected_output = "{'action': 'schedule_mode'}\n"

    request: RequestCall = mocked_responses.requests[("POST", URL(url))][0]
    assert expected_json == request.kwargs["json"]
    assert expected_output == result.output


def test_sessions(mocked_responses: aioresponses, cli_runner: CliRunner) -> None:
    """It exits with a status code of zero."""
    initialise_credential_store(include_account_id=True, include_vin=True)
    fixtures.inject_get_charges(mocked_responses, start="20201101", end="20201130")

    result = cli_runner.invoke(
        __main__.main, "charge sessions --from 2020-11-01 --to 2020-11-30"
    )
    assert result.exit_code == 0, result.exception

    expected_output = (
        "Charge start         Charge end           Duration    Power (kW)  "
        "  Started at    Finished at    Charge gained    Power level    Status\n"
        "-------------------  -------------------  ----------  ------------"
        "  ------------  -------------  ---------------  -------------  --------\n"
        "2020-11-11 01:31:03  2020-11-11 09:30:17  7:59:00     3.10 kW     "
        "  15 %          74 %           59 %             slow           ok\n"
    )
    assert expected_output == result.output


def test_charge_schedule_show(
    mocked_responses: aioresponses, cli_runner: CliRunner
) -> None:
    """It exits with a status code of zero."""
    initialise_credential_store(include_account_id=True, include_vin=True)
    fixtures.inject_get_charging_settings(mocked_responses)

    result = cli_runner.invoke(__main__.main, "charge schedule show")
    assert result.exit_code == 0, result.exception

    expected_output = (
        "Mode: scheduled\n"
        "\n"
        "Schedule ID: 1 [Active]\n"
        "Day        Start time    End time    Duration\n"
        "---------  ------------  ----------  ----------\n"
        "Monday     13:00         13:15       0:15:00\n"
        "Tuesday    05:30         12:30       7:00:00\n"
        "Wednesday  23:30         06:30       7:00:00\n"
        "Thursday   23:00         06:00       7:00:00\n"
        "Friday     13:15         13:30       0:15:00\n"
        "Saturday   13:30         14:00       0:30:00\n"
        "Sunday     13:45         14:30       0:45:00\n"
        "\n"
        "Schedule ID: 2\n"
        "Day        Start time    End time    Duration\n"
        "---------  ------------  ----------  ----------\n"
        "Monday     02:00         02:15       0:15:00\n"
        "Tuesday    03:00         03:30       0:30:00\n"
        "Wednesday  04:00         04:45       0:45:00\n"
        "Thursday   05:00         06:00       1:00:00\n"
        "Friday     06:00         07:15       1:15:00\n"
        "Saturday   07:00         08:30       1:30:00\n"
        "Sunday     08:00         09:45       1:45:00\n"
    )
    assert expected_output == result.output


def test_charging_settings_set(
    mocked_responses: aioresponses, cli_runner: CliRunner
) -> None:
    """It exits with a status code of zero."""
    initialise_credential_store(include_account_id=True, include_vin=True)
    fixtures.inject_get_charging_settings(mocked_responses)
    url = fixtures.inject_set_charge_schedule(mocked_responses, "schedules")

    monday = "--monday clear"
    friday = "--friday T23:30Z,480"
    saturday = "--saturday 19:30,120"
    result = cli_runner.invoke(
        __main__.main, f"charge schedule set 1 {monday} {friday} {saturday}"
    )
    assert result.exit_code == 0, result.exception

    expected_json = {
        "data": {
            "attributes": {
                "schedules": [
                    {
                        "id": 1,
                        "activated": True,
                        "monday": None,
                        "tuesday": {"duration": 420, "startTime": "T04:30Z"},
                        "wednesday": {"duration": 420, "startTime": "T22:30Z"},
                        "thursday": {"duration": 420, "startTime": "T22:00Z"},
                        "friday": {"duration": 480, "startTime": "T23:30Z"},
                        "saturday": {"duration": 120, "startTime": "T18:30Z"},
                        "sunday": {"duration": 45, "startTime": "T12:45Z"},
                    },
                    {
                        "id": 2,
                        "activated": False,
                        "monday": {"startTime": "T01:00Z", "duration": 15},
                        "tuesday": {"startTime": "T02:00Z", "duration": 30},
                        "wednesday": {"startTime": "T03:00Z", "duration": 45},
                        "thursday": {"startTime": "T04:00Z", "duration": 60},
                        "friday": {"startTime": "T05:00Z", "duration": 75},
                        "saturday": {"startTime": "T06:00Z", "duration": 90},
                        "sunday": {"startTime": "T07:00Z", "duration": 105},
                    },
                ]
            },
            "type": "ChargeSchedule",
        }
    }
    expected_output = (
        "{'schedules': [{'id': 1, 'activated': True, "
        "'tuesday': {'startTime': 'T04:30Z', 'duration': 420}, "
        "'wednesday': {'startTime': 'T22:30Z', 'duration': 420}, "
        "'thursday': {'startTime': 'T22:00Z', 'duration': 420}, "
        "'friday': {'startTime': 'T23:30Z', 'duration': 480}, "
        "'saturday': {'startTime': 'T18:30Z', 'duration': 120}, "
        "'sunday': {'startTime': 'T12:45Z', 'duration': 45}}]}\n"
    )
    request: RequestCall = mocked_responses.requests[("POST", URL(url))][0]
    assert expected_json == request.kwargs["json"]
    assert expected_output == result.output


def test_charging_start(mocked_responses: aioresponses, cli_runner: CliRunner) -> None:
    """It exits with a status code of zero."""
    initialise_credential_store(include_account_id=True, include_vin=True)
    url = fixtures.inject_set_charging_start(mocked_responses, "start")

    result = cli_runner.invoke(__main__.main, "charge start")
    assert result.exit_code == 0, result.exception

    expected_json = {
        "data": {
            "attributes": {"action": "start"},
            "type": "ChargingStart",
        }
    }
    expected_output = "{'action': 'start'}\n"
    request: RequestCall = mocked_responses.requests[("POST", URL(url))][0]
    assert expected_json == request.kwargs["json"]
    assert expected_output == result.output
