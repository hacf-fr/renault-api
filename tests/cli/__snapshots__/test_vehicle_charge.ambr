# serializer version: 1
# name: test_charge_history_day
  '''
       Day    Number of charges  Total time charging      Errors
  --------  -------------------  ---------------------  --------
  20201208                    2  8:15:00                       0
  20201205                    1  10:57:00                      0
  
  '''
# ---
# name: test_charge_history_month
  '''
    Month    Number of charges  Total time charging      Errors
  -------  -------------------  ---------------------  --------
   202011                    1  7:59:00                       0
  
  '''
# ---
# name: test_charge_mode_get
  '''
  Charge mode: always
  
  '''
# ---
# name: test_charge_mode_set
  dict({
    'data': dict({
      'attributes': dict({
        'action': 'schedule_mode',
      }),
      'type': 'ChargeMode',
    }),
  })
# ---
# name: test_charge_mode_set.1
  '''
  {'action': 'schedule_mode'}
  
  '''
# ---
# name: test_charge_schedule_show
  '''
  Day        Start time    End time      Duration  Active
  ---------  ------------  ----------  ----------  --------
  Monday     00:00         08:00              480  False
  Tuesday    00:00         08:00              480  False
  Wednesday  00:00         08:00              480  False
  Thursday   00:00         08:00              480  False
  Friday     00:00         08:00              480  False
  Saturday   00:00         08:00              480  False
  Sunday     00:00         08:00              480  False
  
  '''
# ---
# name: test_charge_schedule_show_alternate
  '''
  Mode: Scheduled
  
  Schedule ID: 0 [Active]
  Day        Active
  ---------  --------
  Monday     True
  Tuesday    True
  Wednesday  True
  Thursday   True
  Friday     True
  Saturday   True
  Sunday     True
  
  '''
# ---
# name: test_charging_dacia_start
  dict({
    'data': dict({
      'attributes': dict({
        'action': 'resume',
      }),
      'type': 'ChargePauseResume',
    }),
  })
# ---
# name: test_charging_dacia_start.1
  '''
  {'action': 'resume'}
  
  '''
# ---
# name: test_charging_dacia_stop
  dict({
    'data': dict({
      'attributes': dict({
        'action': 'pause',
      }),
      'type': 'ChargePauseResume',
    }),
  })
# ---
# name: test_charging_dacia_stop.1
  '''
  {'action': 'pause'}
  
  '''
# ---
# name: test_charging_start
  dict({
    'data': dict({
      'attributes': dict({
        'action': 'start',
      }),
      'type': 'ChargingStart',
    }),
  })
# ---
# name: test_charging_start.1
  '''
  {'action': 'start'}
  
  '''
# ---
# name: test_charging_stop
  dict({
    'data': dict({
      'attributes': dict({
        'action': 'stop',
      }),
      'type': 'ChargingStart',
    }),
  })
# ---
# name: test_charging_stop.1
  '''
  {'action': 'stop'}
  
  '''
# ---
# name: test_sessions_40
  '''
  Charge start         Charge end           Duration    Power (kW)    Started at    Finished at    Charge gained    Energy gained    Power level    Status
  -------------------  -------------------  ----------  ------------  ------------  -------------  ---------------  ---------------  -------------  --------
  2020-11-11 01:31:03  2020-11-11 09:30:17  7:59:00     3.10 kW       15 %          74 %           59 %                              slow           ok
  
  '''
# ---
# name: test_sessions_45
  '''
  Charge start         Charge end           Duration    Power (kW)    Started at    Finished at    Charge gained    Energy gained    Power level    Status
  -------------------  -------------------  ----------  ------------  ------------  -------------  ---------------  ---------------  -------------  --------
  2023-04-24 12:12:44  2023-04-24 13:49:39  1:37:00                   43 %          50 %                            4.10000038 kWh                  ok
  2023-04-24 15:19:40  2023-04-24 19:39:27  4:20:00                   49 %          64 %                            9.200001 kWh                    ok
  
  '''
# ---
# name: test_sessions_50
  '''
  Charge start         Charge end           Duration    Power (kW)    Started at    Finished at    Charge gained    Energy gained    Power level    Status
  -------------------  -------------------  ----------  ------------  ------------  -------------  ---------------  ---------------  -------------  --------
  2022-03-16 03:02:16  2022-03-16 05:01:58  1:59:42                   62 %          62 %                                                            error
  2022-03-16 21:18:42  2022-03-16 22:50:08  1:31:26                                 60 %                                                            error
  2022-03-17 02:01:10  2022-03-17 06:02:01  4:00:51                                 74 %                                                            error
  2022-03-18 02:01:13  2022-03-18 06:02:04  4:00:51                                 64 %                                                            error
  2022-03-18 18:34:28  2022-03-18 18:37:57  0:03:29                                 55 %                                                            error
  2022-03-18 18:40:37  2022-03-18 19:17:22  0:36:45                                 89 %                                                            error
  2022-03-19 16:18:18  2022-03-19 17:36:04  1:17:46                                 99 %                                                            error
  2022-03-20 22:21:31  2022-03-21 07:32:50  9:11:19                                 71 %                                                            error
  
  '''
# ---
